File: cloud-driver-platforms/cloud-driver-platforms-desktop/src/desktopMain/kotlin/de/lino/cloud/platform/desktop/utils/PreviewSupport.kt
package de.lino.cloud.platform.desktop.utils

/** Which kind of in-app preview `FilePreviewDialog` renders for a file, based on its content type. */
enum class PreviewKind { TEXT, PDF, DOCX, NONE }

private const val PDF_CONTENT_TYPE = "application/pdf"
private const val DOCX_CONTENT_TYPE = "application/vnd.openxmlformats-officedocument.wordprocessingml.document"

/**
 * Content types (beyond a plain leading `"text/"` prefix) treated as plain-text-previewable - mirrors the
 * text-ish entries of `Constraints.CONTENT_TYPES` (`cloud-driver-api`), reimplemented client-side
 * rather than depended on, same "this module never depends on cloud-driver-api" reasoning
 * [formatBytes] already documents for its own port of `Constraints#resolveBytesToUnit`.
 */
private val TEXT_PREVIEW_CONTENT_TYPES = setOf(
    "application/json", "application/xml", "application/yaml", "application/toml",
)

/** The content type `Constraints.CONTENT_TYPES` (`cloud-driver-api`) maps the `.zip` extension to, server-side - see [isZipArchive]. */
private const val ZIP_CONTENT_TYPE = "application/zip"

/**
 * Whether a file identified by [contentType] is a ZIP archive - double-clicking such a file
 * extracts it into the current folder instead of opening a preview (see
 * `AppViewModel.extractArchive`, `FileBrowserScreen`'s `EntryRow` double-click wiring). Content-type
 * only, not a filename-extension fallback: [contentType] is always server-derived from the file's
 * own extension (`StoredFile#contentType()`, `cloud-driver-api`), so a file actually named `.zip`
 * is guaranteed to carry [ZIP_CONTENT_TYPE] already. Deliberately narrower than [iconFor]'s own
 * archive-icon classification (which also covers 7z/tar/gzip/rar/bzip2/xz for icon purposes) -
 * double-click-to-extract only supports ZIP, the one format the JDK can read without a
 * third-party library.
 */
fun isZipArchive(contentType: String?): Boolean = contentType?.lowercase() == ZIP_CONTENT_TYPE

/**
 * Classifies [contentType] (a file's already-server-assigned content type, e.g.
 * `StoredFileSummaryResponse#contentType()`) into a [PreviewKind] - [PreviewKind.NONE] means
 * `FilePreviewDialog` falls back to a "preview not supported, download to view" message rather
 * than attempting to render anything.
 */
fun previewKindFor(contentType: String?): PreviewKind {
    val type = contentType?.lowercase() ?: return PreviewKind.NONE
    return when {
        type == PDF_CONTENT_TYPE -> PreviewKind.PDF
        type == DOCX_CONTENT_TYPE -> PreviewKind.DOCX
        type.startsWith("text/") || type in TEXT_PREVIEW_CONTENT_TYPES -> PreviewKind.TEXT
        else -> PreviewKind.NONE
    }
}

/**
 * Source-file size ceiling above which `FilePreviewDialog` refuses to even download the file for a
 * PDF/DOCX preview, showing a "too large to preview" message instead - a full in-memory
 * PDFBox/POI parse of a huge file is a real memory/latency cost a preview pane has no business
 * paying; the file is still fully reachable via the existing "Download" action.
 */
const val MAX_PDF_DOCX_PREVIEW_SOURCE_BYTES = 50L * 1024 * 1024

/** Same ceiling as [MAX_PDF_DOCX_PREVIEW_SOURCE_BYTES], for a plain-text file - text decoding is cheap, but the file is still downloaded in full first. */
const val MAX_TEXT_PREVIEW_SOURCE_BYTES = 20L * 1024 * 1024

/** How many bytes of an under-[MAX_TEXT_PREVIEW_SOURCE_BYTES] text file are actually read/rendered - keeps the dialog responsive for e.g. a multi-MB single-line minified file; the rest is never read off disk. */
const val MAX_TEXT_PREVIEW_DISPLAY_BYTES = 2L * 1024 * 1024

File: cloud-driver-platforms/cloud-driver-platforms-mobile/CloudDriverMobile/PreviewSupport.swift
import Foundation

/// Which broad category of preview a file's server-assigned `contentType` supports - originally a
/// strict mirror of cloud-driver-platforms-desktop's `PreviewSupport.kt#PreviewKind` (text/PDF/DOCX
/// only), extended 2026-09-06 with `.image` (JPG/JPEG/PNG), per Lino's own request - a mobile-only
/// addition for now, not ported back to the desktop client. Unlike the desktop app (no JVM
/// equivalent to QuickLook exists, so it hand-rolls PDF/DOCX rendering via PDFBox/POI, and would
/// need a third library just for images), this app doesn't need its own renderer per kind -
/// `QLPreviewController` (`FilePreviewView.swift`) already knows how to display all four natively
/// from the downloaded file's own extension - `.image` needed no changes to that view at all, only
/// to this file's own classification/size-cap logic. This enum still exists to gate *which* files
/// are even offered a single-tap preview at all, and to pick the right size cap before downloading
/// anything.
enum PreviewKind {
    case text
    case pdf
    case docx
    case image
    case none
}

/// Content types QuickLook can't infer are text from a `text/` prefix alone - matches the desktop
/// app's own `previewKindFor`'s additional text-ish grouping (json/xml/yaml/toml), reimplemented
/// here rather than shared code, the same "hand-kept-in-sync port" convention this module's own
/// `ByteFormat.swift`/`FileIcons.swift` already document for their server-mirrored logic.
private let additionalTextPreviewContentTypes: Set<String> = [
    "application/json", "application/xml", "application/x-yaml", "text/yaml",
    "application/x-yaml", "application/toml"
]

/// The exact content types `Constraints.CONTENT_TYPES` (`cloud-driver-api`) assigns `.jpg`/`.jpeg`/
/// `.png` - **JPG/JPEG/PNG only, per Lino's own explicit request**, not a broad `image/` prefix
/// match the way text uses one: an unlisted image format (GIF, HEIC, WEBP, ...) falls through to
/// `.none` below just like any other unsupported type, the same "flag it, don't scope-creep"
/// precedent this codebase applies elsewhere (e.g. archive extraction being ZIP-only, not
/// "zip/rar/7z").
private let imagePreviewContentTypes: Set<String> = ["image/jpeg", "image/png"]

/// 20 MB - matches `cloud-driver-platforms-desktop`'s `MAX_TEXT_PREVIEW_SOURCE_BYTES`.
let maxTextPreviewSourceBytes: Int64 = 20 * 1024 * 1024
/// 50 MB - matches `cloud-driver-platforms-desktop`'s `MAX_PDF_DOCX_PREVIEW_SOURCE_BYTES`.
let maxPdfDocxPreviewSourceBytes: Int64 = 50 * 1024 * 1024
/// 50 MB - same ceiling as PDF/DOCX rather than a smaller, image-specific one: a JPG/PNG is
/// typically far smaller than that in practice, but a high-resolution photo/screenshot can still
/// legitimately approach it, and QuickLook renders an image without this app decoding it into
/// memory itself first (unlike `rememberThumbnail`'s own row-icon path on the desktop client),
/// so there's no extra memory-pressure reason here to set a tighter cap than the other kinds.
let maxImagePreviewSourceBytes: Int64 = 50 * 1024 * 1024

func previewKind(for contentType: String) -> PreviewKind {
    if contentType.hasPrefix("text/") || additionalTextPreviewContentTypes.contains(contentType) {
        return .text
    }
    if contentType == "application/pdf" {
        return .pdf
    }
    if contentType == "application/vnd.openxmlformats-officedocument.wordprocessingml.document" {
        return .docx
    }
    if imagePreviewContentTypes.contains(contentType) {
        return .image
    }
    return .none
}

/// The size cap that applies to `kind` - `nil` for `.none`, since there's nothing to size-check
/// for a file this app never offers to preview in the first place.
func maxPreviewSourceBytes(for kind: PreviewKind) -> Int64? {
    switch kind {
    case .text: return maxTextPreviewSourceBytes
    case .pdf, .docx: return maxPdfDocxPreviewSourceBytes
    case .image: return maxImagePreviewSourceBytes
    case .none: return nil
    }
}
